sw/host/opentitanlib/src/transport/cw310/mod.rs

```diff
@@ -115,6 +115,19 @@ impl CW310 {
                 }
                 false
             });
+
+            // Some systems will erroneously enumerate the CW310's UARTs as PCI devices.
+            // See <https://github.com/lowRISC/opentitan/issues/18491>.
+            if ports.is_empty() {
+                log::warn!("Detected zero CW310 USB devices. Guessing device names.");
+                ports = serialport::available_ports()?;
+                ports.retain(|port| {
+                    const UART0_NAME_GUESS: &str = "/dev/ttyACM0";
```

Comment:
Hm... are you sure you want to be taking over a user's random devices unprompted? You don't know which devices these could be. ;)

Instead, do we have a way for users to pass configuration for test setup? Maybe something with bazelrc and an opentitantool config?

Reply:
Hmm, I'd prefer for it to be automatic.

WDYT about checking `/sys/class/tty/${device_name}/device/uevent` before proceeding with the guessed device? If we check that the PRODUCT line begins with `2b3e/c310/`, that should indicate it's actually the CW310 device, right?

```console
$ cat  /sys/class/tty/ttyACM0/device/uevent
DEVTYPE=usb_interface
DRIVER=cdc_acm
PRODUCT=2b3e/c310/900
TYPE=0/0/0
INTERFACE=2/2/0
MODALIAS=usb:v2B3EpC310d0900dc00dsc00dp00ic02isc02ip00in01
```

Reply:
I'm not the biggest fan of maintaining workarounds for bugs in other OSes, but this isn't the end of the world. Hopefully one of the other reviewers can lean more to one side or the other, heh.

For actual implementation, I'd recommend using the udev crate and probing for more of the keys. We have ID_USB_MODEL, for example, and the serial port crate doesn't dive that deep. Using the library is probably better than trying to navigate the sysfs object hierarchy directly (assuming it is even mounted where we think).

Reply:
This isn't our bug and I don't really want to add a bunch of extra stuff of opentitantool just because udev happens to have a broken configuration.

I'd be more sympathetic to adding a patch to the serialport-rs crate that jazzes up the [`port_type`](https://github.com/serialport/serialport-rs/blob/main/src/posix/enumerate.rs#L73-L92) function to catch the broken udev properties and returns the correct struct (see //third_party/rust/BUILD to see how to apply patches to crates).  We can then just drop the patch when udev gets fixed.

_Or_, we might realize that the ports on Dan's system get mapped exactly the same way as things get mapped in the CI execution container at lowrisc, and affected users could just throw `--define cw310=lowrisc` on their command line or into their `.bazelrc-site` until such time as udev gets fixed.

Reply:
> I'd be more sympathetic to adding a patch to the serialport-rs crate that jazzes up the [port_type](https://github.com/serialport/serialport-rs/blob/main/src/posix/enumerate.rs#L73-L92) function to catch the broken udev properties and returns the correct struct (see //third_party/rust/BUILD to see how to apply patches to crates). We can then just drop the patch when udev gets fixed.

Oh, interesting approach. I'll look into that.

> Or, we might realize that the ports on Dan's system get mapped exactly the same way as things get mapped in the CI execution container at lowrisc, and affected users could just throw --define cw310=lowrisc on their command line or into their .bazelrc-site until such time as udev gets fixed.

Unfortunately, this is not the case. That configuration uses `--cw310-uarts=/dev/ttyACM_CW310_{0,1}`, but my system has `/dev/ttyACM{0,1}`. I guess I might be able to use the SYMLINK operation in my udev rules.

Reply:
I've created a PR that patches the serialport crate: https://github.com/lowRISC/opentitan/pull/18656